### PR TITLE
Arm64: Memory barrier improvements

### PR DESCRIPTION
**Pull request by kunalspathak**
- Generate store barriers wherever possible. Currently, we generate full barriers for stores.
- Generate one-way barriers for `volatile` variable which makes the speed of `volatile` declared variables as observed in https://github.com/dotnet/runtime/issues/60232. Thanks @EgorBo  for suggesting the solution as well.

### CODE REVIEW

**Comment by alexrp** [comment]
I haven't reviewed this PR in detail, but I would just advise caution when it comes to memory barriers on ARM64:

* https://github.com/mono/mono/pull/3418
* https://github.com/mono/mono/pull/3464

Basically, some of the instructions don't quite have the semantics you might expect.

**Comment by kunalspathak** [comment]
I see approx. 0.58% and 0.28% improvement in RPS in mvc and json benchmarks respectively. I believe these are in error range.

**Comment by kunalspathak** [comment]
> I haven't reviewed this PR in detail, but I would just advise caution when it comes to memory barriers on ARM64:

Thanks @alexrp . This PR just extends the optimal memory barriers for `volatile` keyword and for one scenario where we can use `dmb ishst`.

**Comment by kunalspathak** [comment]
@dotnet/jit-contrib

**Comment by EgorBo** [comment]
LGTM, @VSadov could you please take a quick look if you have time
tldr:
```
volatileVariable = 42;
```
used to emit a full memory barrier, now it emits a store-only one.

Also, for most cases stores/loads for variables marked as "volatile" now actually don't emit memory barriers at all and use e.g. `stlr` instead in case of store, basically the same what Volatile.Write emits today, see https://github.com/dotnet/runtime/issues/60232

**Review by EgorBo — APPROVED** [review]

File: src/coreclr/jit/codegenarm64.cpp [hunk 1]
```diff
@@ -3280,7 +3280,7 @@ void CodeGen::genCodeForStoreInd(GenTreeStoreInd* tree)
             else
             {
                 // issue a full memory barrier before a volatile StInd
-                instGen_MemoryBarrier();
+                instGen_MemoryBarrier(BARRIER_STORE_ONLY);
```
Comment (VSadov) [inline src/coreclr/jit/codegenarm64.cpp:3283]: What is the typical scenario when this branch is taken? When value is a struct?

Comment (VSadov) [reply src/coreclr/jit/codegenarm64.cpp:3283]: As I understand the purpose of this barrier is to have release semantics for storing into volatile variable that does not fit into a single register (otherwise stlr could be used). 

I think this needs a full barrier, since unlike `stlr`, `dmb ishst` does not wait for loads in progress.

Comment (VSadov) [reply src/coreclr/jit/codegenarm64.cpp:3283]: Basically, `ldar` can be replaced with `ldr; dmb ishld` , but there is no such equivalency between `stlr` and `dmb ishst; str` because ishst is too weak.

Comment (kunalspathak) [reply src/coreclr/jit/codegenarm64.cpp:3283]: From https://developer.arm.com/documentation/100941/0100/Barriers, trying to understand what you stated.

image

image

Basically, for `ishst`, loads can still be reordered around barrier and hence it is weaker than the `ishld` where loads/stores need to wait till the barrier is complete.

If we change from full barrier to `ishst`, we might have a load that should have been completed but got reordered and might end up reading the wrong value (pre-updated value). Is my understanding correct?

Comment (VSadov) [reply src/coreclr/jit/codegenarm64.cpp:3283]: Yes, loads that appear after ishst, in program order, may speculatively happen ahead of the store.

Comment (VSadov) [reply src/coreclr/jit/codegenarm64.cpp:3283]:
```cs
static volatile int x;
static volatile int y;

static int xx;
static int yy;

---- one thread:
x = 42;
yy = y;

--- another thread:
y = 42;
xx = x;

```

Can both `xx` and `yy` end up 0 ?

Comment (kunalspathak) [reply src/coreclr/jit/codegenarm64.cpp:3283]: Makes sense. I will revert the change related to `shst` then.

**Review by VSadov — APPROVED** [review]
LGTM!

**Comment by EgorBo** [comment]
Improvement on ubuntu-arm64: https://github.com/dotnet/perf-autofiling-issues/issues/2981
